SalesforceCommerceCloud/pwa-kit
# [Performance] Share connections to backend systems between SSR requests

GUS: https://gus.lightning.force.com/lightning/r/ADM_Work__c/a07EE00000BOR5gYAH/view

# Description

The idea behind this PR is to re-use a single TCP connection to send and receive multiple HTTP requests/responses server-side, as opposed to opening a new connection for every single request/response pair.

We currently use the SDK proxy to send server-side requests to the Commerce API.

This PR adds an Agent with keepAlive support to the Commerce SDK isomorphic configuration to re-use the connections server-side.



### __Docs__ 

- `commerce-sdk-isomorphic` configuration: https://github.com/SalesforceCommerceCloud/commerce-sdk-isomorphic#advanced-options
- `node-fetch` with a custom Agent that supports keepAlive: https://github.com/node-fetch/node-fetch#custom-agent
- Node.js Agent: https://nodejs.org/api/http.html#new-agentoptions








# Types of Changes




> Breaking changes include:
>
> - Removing a public function or component or prop
> - Adding a required argument to a function
> - Changing the data type of a function parameter or return value
> - Adding a new peer dependency to `package.json`

# Changes

- Added a new Node.js Agent with keepAlive support to the `commerce-sdk-isomorphic` configuration.

# How to Test-Drive This PR

Follow the steps detailed in John's comment:
https://github.com/SalesforceCommerceCloud/pwa-kit/pull/244#issuecomment-989331030


Verify that we see only one `localhost:3000` connection on the consecutive curls request made within the 8sec interval of that last curl.

![one-localhost](https://user-images.githubusercontent.com/6024102/146041968-494972df-9f2f-4e9f-af41-017b72736dec.gif)

# Known Issues

- After starting the dev server for the first time the connection isn't reused for the first request, subsequent requests within the timeout will reuse the connection as intended.

# Checklists




## General


## Accessibility Compliance

You must check off all items in **one** of the follow two lists:


_or..._


## Localization
